### PR TITLE
Jupyter

### PR DESCRIPTION
**Pull request by felipeZ**
Create jupyter notebooks for the tutorials:
* [x] Gromacs QMMM 
* [x] LAMMPS QMMM 
* [x] DFT-GWBSE 
* [x] KMC

### CODE REVIEW

File: tools/dftgwbse_CH4/DFTGWBSE_ENERGY.ipynb [hunk 1]
```diff
@@ -0,0 +1,759 @@
...
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "==================================================\r\n",
+      "========   VOTCA (http://www.votca.org)   ========\r\n",
+      "==================================================\r\n",
+      "\r\n",
+      "please submit bugs to bugs@votca.org\r\n",
+      "\r\n",
+      "xtp_tools, version 1.7-dev gitid: 1c4a17f (dirty) (compiled May 28 2020, 15:47:26)\r\n",
```
Comment (junghans) [inline tools/dftgwbse_CH4/DFTGWBSE_ENERGY.ipynb:42]: Is there a way to not store all these outputs?

Comment (felipeZ) [reply tools/dftgwbse_CH4/DFTGWBSE_ENERGY.ipynb:42]: Yes, I am going to redirect the output to a local file, so we just show the command.

File: LAMMPS/KMC_Thiophene/LAMMPS_KMC.ipynb [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+     "output_type": "stream",
+     "text": [
+      "The option 'resolution_sites' on file 'OPTIONFILES/eanalyze.xml' has been set to '0.03'\n",
+      "==================================================\n",
+      "========   VOTCA (http://www.votca.org)   ========\n",
+      "==================================================\n",
+      "\n",
+      "please submit bugs to bugs@votca.org\n",
+      "\n",
+      "xtp_run, version 1.7-dev gitid: 6e47b87 (dirty) (compiled Jun  2 2020, 10:18:25)\n",
```
Comment (junghans) [inline LAMMPS/KMC_Thiophene/LAMMPS_KMC.ipynb:194]: Output is still there.

Comment (felipeZ) [reply LAMMPS/KMC_Thiophene/LAMMPS_KMC.ipynb:194]: Shall I remove all the output from all the jupyter cell? I was thing that we only one to remove long outputs and leave the short ones

File: tools/dftgwbse_CH4/DFTGWBSE_ENERGY.ipynb (hunk 1 above)
Comment (felipeZ) [reply tools/dftgwbse_CH4/DFTGWBSE_ENERGY.ipynb]: I mean the output of the cell is automatically used to generate the documentation. Maybe we want to leave some output so the user have an idea what can of output to expect from the program

File: LAMMPS/KMC_Thiophene/LAMMPS_KMC.ipynb (hunk 2 above)
Comment (junghans) [reply LAMMPS/KMC_Thiophene/LAMMPS_KMC.ipynb]: @JensWehner what do you think, I just don't like the idea of having hard-coded git hashes in here, which change all the time.

Comment (JensWehner) [reply LAMMPS/KMC_Thiophene/LAMMPS_KMC.ipynb]: I agree with Christoph, nothing should be in the repo which can be created by just running it unless it really takes a long time. Can we not create the documentation from the notebook by executing it?

Comment (felipeZ) [reply LAMMPS/KMC_Thiophene/LAMMPS_KMC.ipynb]: OK, I am cleaning the cell of the jupyter notebooks and we run them in the CI

**Comment by felipeZ** [comment]
@JensWehner  the Notebooks are ready, they take considerable time to run and therefore running then in the CI is not feasible, what shall I do then?

**Comment by felipeZ** [comment]
@junghans  I have created [this CMakeLists file]() for running the Jupyter notebooks and adding them to the Sphinx folder. For running the notebooks I need to have loaded the `VOTCARC.bash` file to add the binaries in the `PATH`. I have seen that you had done similar in `csg`, could you please give me some advice on how to proceed?

**Comment by junghans** [comment]
> @junghans I have created this CMakeLists file for running the Jupyter notebooks and adding them to the Sphinx folder. For running the notebooks I need to have loaded the `VOTCARC.bash` file to add the binaries in the `PATH`. I have seen that you had done similar in `csg`, could you please give me some advice on how to proceed?

The VOTCARC.bash contains the install paths no the path where stuff is build.

For csg binaries you can do:
```cmake
get_target_property(VOTCA_PATH VOTCA::csg_call VOTCA_PATH)
...
 add_custom_command(OUTPUT ${VOTCA_SPHINX_DIR}/${_FILE}
    ${CMAKE_COMMAND} -E env PATH=PATH=${VOTCA_PATH}:$ENV{PATH} COMMAND ${JUPYTER_EXECUTABLE} nbconvert --to notebook --ExecutePreprocessor.timeout=1000 --execute ${_notebook} --output ${VOTCA_SPHINX_DIR}/${_FILE}
    )
```

If that works for csg, we need to implement something similar for xtp.

**Comment by felipeZ** [comment]
@junghans  I have set a [XTPBIN property](https://github.com/votca/xtp/blob/tutorial/src/tools/CMakeLists.txt#L5) so now the notebook can find the binaries, see the [custom command](https://github.com/votca/xtp-tutorials/blob/jupyter/share/doc/CMakeLists.txt#L17). Now, I need a way to add the path to the XML files of XTP (which are install in `${VOTCASHARE}/xtp/xml`)

**Comment by junghans** [comment]
> @junghans I have set a [XTPBIN property](https://github.com/votca/xtp/blob/tutorial/src/tools/CMakeLists.txt#L5) so now the notebook can find the binaries, see the [custom command](https://github.com/votca/xtp-tutorials/blob/jupyter/share/doc/CMakeLists.txt#L17). Now, I need a way to add the path to the XML files of XTP (which are install in `${VOTCASHARE}/xtp/xml`)

I would name it `XTP_PATH` instead (`XTPBIN` is bit strange, @JensWehner any comment) and make it a property of libvotca_xtp instead of every tool.
Similarly you could add a property like XTP_XML_PATH to `libvotca_xtp` and then set the environment variable `VOTCA_SHARE` (?) to that value here.

**Comment by felipeZ** [comment]
> > @junghans I have set a [XTPBIN property](https://github.com/votca/xtp/blob/tutorial/src/tools/CMakeLists.txt#L5) so now the notebook can find the binaries, see the [custom command](https://github.com/votca/xtp-tutorials/blob/jupyter/share/doc/CMakeLists.txt#L17). Now, I need a way to add the path to the XML files of XTP (which are install in `${VOTCASHARE}/xtp/xml`)
> 
> I would name it `XTP_PATH` instead (`XTPBIN` is bit strange, @JensWehner any comment) and make it a property of libvotca_xtp instead of every tool.
> Similarly you could add a property like XTP_XML_PATH to `libvotca_xtp` and then set the environment variable `VOTCA_SHARE` (?) to that value here.

Thanks for the advice! I would do so.

**Comment by JensWehner** [comment]
I agree with  @junghans `XTPBIN` is not a good name. I like the rest.

**Comment by felipeZ** [comment]
@JensWehner  This PR is ready but It doesn't work because It needs the changes in the XTP `tutorial` branch

File: tools/dftgwbse_CO_geoopt/CO.xyz [hunk 3]
```diff
@@ -1,4 +1,4 @@
 2
-CO man
-C 0.0 0.0 0.0
-O 1.2 0.0 0.0
+CO
+C -0.0254018 -6.69351e-08 8.88366e-06
+O 1.2254 6.69351e-08 -8.88366e-06
```
Comment (JensWehner) [inline tools/dftgwbse_CO_geoopt/CO.xyz:4]: Why the change?

Comment (felipeZ) [reply tools/dftgwbse_CO_geoopt/CO.xyz:4]: I want to used the optimized geometry to reduce cycles. But anyway, I reduce the optimization cycles to 1. Shall I go back to the previous geometry?

Comment (JensWehner) [reply tools/dftgwbse_CO_geoopt/CO.xyz]: Does it take that long, I wanted to have 2-3 steps so that we see sth.

**Comment by felipeZ** [comment]
@junghans do you have an idea why the CI is still using the old XTP configuration? I have moved the `XTP-TUTORIAL.rst` file from the XTP repo to this one.  But the GH Actions reports that:
```
 CMake Error at share/doc/CMakeLists.txt:29 (add_custom_command):
  Attempt to add a custom rule to output

    /__w/xtp-tutorials/xtp-tutorials/builddir/sphinx/XTP-TUTORIAL.rst.rule

  which already has a custom rule.
```

**Comment by junghans** [comment]
> @junghans do you have an idea why the CI is still using the old XTP configuration? I have moved the `XTP-TUTORIAL.rst` file from the XTP repo to this one. But the GH Actions reports that:
> 
> ```
>  CMake Error at share/doc/CMakeLists.txt:29 (add_custom_command):
>   Attempt to add a custom rule to output
> 
>     /__w/xtp-tutorials/xtp-tutorials/builddir/sphinx/XTP-TUTORIAL.rst.rule
> 
>   which already has a custom rule.
> ```

Have you merged xtp's master into votca/votca's master yet?

**Comment by felipeZ** [comment]
> > @junghans do you have an idea why the CI is still using the old XTP configuration? I have moved the `XTP-TUTORIAL.rst` file from the XTP repo to this one. But the GH Actions reports that:
> > ```
> >  CMake Error at share/doc/CMakeLists.txt:29 (add_custom_command):
> >   Attempt to add a custom rule to output
> > 
> >     /__w/xtp-tutorials/xtp-tutorials/builddir/sphinx/XTP-TUTORIAL.rst.rule
> > 
> >   which already has a custom rule.
> > ```
> 
> Have you merged xtp's master into votca/votca's master yet?

Right! I forgot to do so

**Comment by felipeZ** [comment]
We need to solve first [votca/xtp#475](https://github.com/votca/xtp/issues/475)

**Comment by felipeZ** [comment]
@junghans  do you know why `XTP-Tutorials` cannot find the `votca_xtp` target:
```bash
 CMake Error at share/doc/CMakeLists.txt:17 (get_target_property):
  get_target_property() called with non-existent target "votca_xtp".
 ```

**Comment by junghans** [comment]
I see that is in the module build and it fails because we haven’t exported the votca_xtp target yet.

Let me have look at it later. You should be able to reproduce this with `-DBUILD_MODULE=On`.

File: share/doc/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+endif()
+
+file(GLOB NOTEBOOKS "${PROJECT_SOURCE_DIR}/GROMACS/Methane/QMMM_GROMACS.ipynb")
+list(APPEND NOTEBOOKS "${PROJECT_SOURCE_DIR}/LAMMPS/Thiophene/QMMM_LAMMPS.ipynb")
+list(APPEND NOTEBOOKS "${PROJECT_SOURCE_DIR}/tools/dftgwbse_CH4/DFTGWBSE_ENERGY.ipynb")
+list(APPEND NOTEBOOKS "${PROJECT_SOURCE_DIR}/tools/dftgwbse_CO_geoopt/DFTGWBSE_OPTIMIZATION.ipynb")
+# list(APPEND NOTEBOOKS "${PROJECT_SOURCE_DIR}/GROMACS/KMC_Methane/GROMACS_KMC.ipynb")
+list(APPEND NOTEBOOKS "${PROJECT_SOURCE_DIR}/LAMMPS/KMC_Thiophene/LAMMPS_KMC.ipynb")
+
+find_program(JUPYTER_EXECUTABLE NAMES jupyter "Interactive computing environment (https://jupyter.org/)")
```
Comment (junghans) [inline share/doc/CMakeLists.txt:12]: I think there is a `DOC` missing before the documentation.

File: share/doc/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+find_program(PIP_EXECUTABLE NAMES pip "Python package installer (https://pip.pypa.io/en/stable/)")
```
Comment (junghans) [inline share/doc/CMakeLists.txt:13]: We need to required this.

File: share/doc/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+add_custom_target(jupyter_client COMMAND ${PIP_EXECUTABLE} install wheel ipykernel jupyter_client sphinx>=2.1 nbsphinx --user)
```
Comment (junghans) [inline share/doc/CMakeLists.txt:15]: I really don't like the idea that we are automatically installing some crap in the user home directory.

We should just check if the required python modules are there (see the check for cma in csg-tutorials) and then fail/ don't build anything if they are don't. We can leave the pip install as a custom command, but the user has to run it manually.

Comment (felipeZ) [reply share/doc/CMakeLists.txt:15]: I am going to remove the pip command for the moment being and I am going to follow your advice about checking the required python modules.

**Comment by felipeZ** [comment]
@junghans  I am having issues compiling this branch with the new CMAKE configuration:
```
CMake Error at CMakeLists.txt:106 (_find_package):
  Could not find a package configuration file provided by "VOTCA_XTP" with
  any of the following names:

    VOTCA_XTPConfig.cmake
    votca_xtp-config.cmake

  Add the installation prefix of "VOTCA_XTP" to CMAKE_PREFIX_PATH or set
  "VOTCA_XTP_DIR" to a directory containing one of the above files.  If
  "VOTCA_XTP" provides a separate development package or SDK, be sure it has
  been installed.
Call Stack (most recent call first):
  xtp-tutorials/CMakeLists.txt:11 (find_package)
```
I cannot provide a `VOTCA_XTP_DIR` because I am installing the whole VOTCA library from scratch

**Comment by junghans** [comment]
@felipeZ https://github.com/votca/votca/pull/346

File: CMakeModules/FindJUPYTER_NBCONVERT.cmake [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+find_program(JUPYTER_EXECUTABLE NAMES jupyter DOC "Interactive computing environment (https://jupyter.org/)")
+find_package_handle_standard_args(JUPYTER REQUIRED_VARS JUPYTER_EXECUTABLE)
+
+set(JUPYTER_NBCONVERT_FOUND FALSE)
+if(JUPYTER_FOUND)
+  execute_process(COMMAND ${PYTHON_EXECUTABLE} -c "import cma"
```
Comment (junghans) [inline CMakeModules/FindJUPYTER_NBCONVERT.cmake:27]: `cma`?

Comment (felipeZ) [reply CMakeModules/FindJUPYTER_NBCONVERT.cmake:27]: sorry, I forgot to change the test

**Comment by junghans** [comment]
Nice, is there a way to check for the version number as well?

(Have look at `FindCLANG_FORMAT.cmake` in `votca/votca` for an example on output parsing.)

File: CMakeModules/FindJUPYTER_NBCONVERT.cmake [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    OUTPUT_VARIABLE nbconvert_version
+    ERROR_QUIET OUTPUT_STRIP_TRAILING_WHITESPACE)
+  if(${nbconvert_version} MATCHES "^([0-9]+)\\.([0-9]+)\\.([0-9]+)$")
+    set(JUPYTER_NBCONVERT_VERSION ${nbconvert_version})
+  else()
+    set(JUPYTER_NBCONVERT_VERSION 0.0)
+  endif()
+endif()
+
+find_package_handle_standard_args(JUPYTER_NBCONVERT REQUIRED_VARS JUPYTER_EXECUTABLE JUPYTER_NBCONVERT_VERSION)
```
Comment (junghans) [inline CMakeModules/FindJUPYTER_NBCONVERT.cmake:36]:
```suggestion
find_package_handle_standard_args(JUPYTER_NBCONVERT REQUIRED_VARS JUPYTER_EXECUTABLE VERSION_VAR JUPYTER_NBCONVERT_VERSION)
```

**Comment by felipeZ** [comment]
@JensWehner @junghans  How can I know if the Jupyter notebooks are running correctly? can we merge this one?

**Comment by felipeZ** [comment]
> @JensWehner @junghans How can I know if the Jupyter notebooks are running correctly?

**Comment by felipeZ** [comment]
I just got confused with the CI. Now I see it

**Comment by felipeZ** [comment]
@junghans  can I skip the Sphinx building in ubuntu?

File: share/doc/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+endif()
+
+file(GLOB NOTEBOOKS "${PROJECT_SOURCE_DIR}/GROMACS/Methane/QMMM_GROMACS.ipynb")
+list(APPEND NOTEBOOKS "${PROJECT_SOURCE_DIR}/LAMMPS/Thiophene/QMMM_LAMMPS.ipynb")
+list(APPEND NOTEBOOKS "${PROJECT_SOURCE_DIR}/tools/dftgwbse_CH4/DFTGWBSE_ENERGY.ipynb")
+list(APPEND NOTEBOOKS "${PROJECT_SOURCE_DIR}/tools/dftgwbse_CO_geoopt/DFTGWBSE_OPTIMIZATION.ipynb")
+# list(APPEND NOTEBOOKS "${PROJECT_SOURCE_DIR}/GROMACS/KMC_Methane/GROMACS_KMC.ipynb")
+list(APPEND NOTEBOOKS "${PROJECT_SOURCE_DIR}/LAMMPS/KMC_Thiophene/LAMMPS_KMC.ipynb")
+
+find_package(JUPYTER_NBCONVERT 5.4.1 REQUIRED)
```
Comment (junghans) [inline share/doc/CMakeLists.txt:12]: I think the previous solution was better now the build requires `nbconvert` even to build xtp without any documentation.

Comment (felipeZ) [reply share/doc/CMakeLists.txt]: I have made a `revert`. The problem is that sphinx needs `nbconvert` to create the documentation and there is not suitable `nbconvert` for ubuntu. So, can we skip the Sphinx build on ubuntu

Comment (JensWehner) [reply share/doc/CMakeLists.txt]: We only have to use it on one single platform. Despite all that, @felipeZ could you add sth. to the developer guide in votca/votca about what the documentation does?

Comment (junghans) [reply share/doc/CMakeLists.txt]: > I have made a `revert`. The problem is that sphinx needs `nbconvert` to create the documentation and there is not suitable `nbconvert` for ubuntu. So, can we skip the Sphinx build on ubuntu

Wouldn't
```
find_package(JUPYTER_NBCONVERT 5.4.1)
if(JUPYTER_NBCONVERT_FOUND AND SPHINX_FOUND)
```
take care of that?

Comment (junghans) [reply share/doc/CMakeLists.txt]: > We only have to use it on one single platform. Despite all that, @felipeZ could you add sth. to the developer guide in votca/votca about what the documentation does?

Correct. We can exclude stuff in the setup action, but I don't want the user to not being able to build xtp on Ubuntu because `JUPYTER_NBCONVERT` is required. We should try to find a new enough version and if that fails, we just skip that part of the documentation build.

Comment (felipeZ) [reply share/doc/CMakeLists.txt]: The problem is the following, there is a `index.rst` file in `VOTCA/VOTCA` that contains a reference to the `XTP-TUTORIAL.rst` file. When sphinx  is trying to build the documentation it will go first to `index.rst` then to `XTP-TUTORIAL.rst`. The problem is that inside that last file there are reference to the jupyter notebooks and sphinx will always try to convert those Jupyter notebooks into documentation by running them.
On the other hand,  the statement:
```cmake
find_package(JUPYTER_NBCONVERT 5.4.1)
if(JUPYTER_NBCONVERT_FOUND AND SPHINX_FOUND)
```
Can just try to run the notebooks or just copy them as they are, but Sphinx will always try to run them anyway and seems there is no  `nbconvert` tool it would fail.
So, I agree with @junghans  that the best alternative is just excluding the Sphinx build on Ubuntu.

Comment (junghans) [reply share/doc/CMakeLists.txt]: If we exclude it in the CI it will still trip up user on ubuntu running it manually.

Can we tell Sphinx to ignore the broken links? Or just create a mock version of xtp-tutorials.rst?

Comment (felipeZ) [reply share/doc/CMakeLists.txt]: We can ignore broken links, with `nbsphinx_allow_errors = True`  in the `config.py` file. I am going to do a PR

**Comment by junghans** [comment]
On openSUSE: `ModuleNotFoundError: No module named 'pandas'`

On Fedora Rawhide: `ModuleNotFoundError: No module named 'seaborn'`

**Comment by junghans** [comment]
On openSUSE:
```
Notebook error:
PandocMissing in DFTGWBSE_ENERGY.ipynb:
Pandoc wasn't found.
Please check that pandoc is installed:
http://pandoc.org/installing.html
```

**Comment by felipeZ** [comment]
@junghans  @JensWehner  It is working!!

**Comment by JensWehner** [comment]
@felipeZ good work.

**Comment by junghans** [comment]
Fixup in #26
